HARK/distribution.py

```diff
@@ -930,29 +928,204 @@ def draw(self, N, X=None, exact_match=False):
+            The the expectation is computed at f(dstn, *args).
+
+        Returns
+        -------
+        f_exp : np.array or scalar
+            The expectation of the function at the queried values.
+            Scalar if only one value.
+        """
+
+        if func is None:
+            # if no function is provided, it's much faster to go straight
+            # to dot product instead of calling the dummy function.
+            f_query = self.X
+        else:
+            # if a function is provided, we need to add one more dimension,
+            # the nature dimension, to any inputs that are n-dim arrays.
+            # This allows numpy to easily broadcast the function's output.
+            args = [
+                arg[..., np.newaxis] if isinstance(arg, np.ndarray) else arg
+                for arg in args
+            ]
+
+            f_query = func(self.X, *args)
+
+        f_exp = np.dot(f_query, self.pmf)
+
+        return f_exp
+
+    def distr_of_function(self, func=lambda x: x, *args):
+        """
+        Finds the distribution of a random variable Y that is a function
+        of discrete random variable X, Y=f(X).
+
+        Parameters
+        ----------
+        func : function
+            The function to be evaluated.
+            This function should take an array of shape dstn.dim().
+            It may also take other arguments *args.
+        *args :
+            Additional non-stochastic arguments for func,
+            The function is computed at f(dstn, *args).
+
+        Returns
+        -------
+        f_dstn : DiscreteDistribution
+            The distribution of func(dstn).
+        """
+        # we need to add one more dimension,
+        # the nature dimension, to any inputs that are n-dim arrays.
+        # This allows numpy to easily broadcast the function's output.
+        args = [
+            arg[..., np.newaxis] if isinstance(arg, np.ndarray) else arg for arg in args
+        ]
+        f_query = func(self.X, *args)
+
+        f_dstn = DiscreteDistribution(list(self.pmf), f_query, seed=self.seed)
+
+        return f_dstn
+
+
+class DiscreteDistributionXRA(DiscreteDistribution):
+    def __init__(
+        self,
+        pmf,
+        data,
+        seed=0,
+        coords=None,
+        dims=None,
+        name="DiscreteDistributionXRA",
+        attrs=None,
+    ):
+
+        if data.ndim < 2:
+            data = data[np.newaxis, ...]
+
+        if attrs is None:
+            attrs = {}
+
+        attrs["pmf"] = np.asarray(pmf)
+        attrs["seed"] = seed
+        attrs["RNG"] = np.random.RandomState(seed)
+
+        self._xarray = DataArray(
+            data=data,
+            coords=coords,
+            dims=dims,
+            name=name,
+            attrs=attrs,
+        )
+
+    @property
+    def xarray(self):
+        """
+        Returns the underlying xarray.DataArray object.
+        """
+        return self._xarray
+
+    @property
+    def X(self):
```

Comment:
Reminds me of https://github.com/econ-ark/HARK/issues/1051

Reply:
Happy to move on from `X`. `xarray`s already use the values and data attribute to return the underlying rvs so this would be easy to do. This was just a patch but we can deprecate the use of `X`